anta/reporter/csv_reporter.py

```diff
@@ -0,0 +1,97 @@
+# Copyright (c) 2023-2024 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+"""CSV Report management for ANTA."""
+
+# pylint: disable = too-few-public-methods
+from __future__ import annotations
+
+import csv
+import logging
+import pathlib
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from anta.result_manager import ResultManager
+    from anta.result_manager.models import TestResult
+
+logger = logging.getLogger(__name__)
+
+
+class ReportCsv:
+    """Build a CSV report."""
+
+    @dataclass()
+    class Headers:
+        """Headers for the CSV report."""
+
+        device: str = "Device"
+        test_name: str = "Test Name"
+        test_status: str = "Test Status"
+        messages: str = "Message(s)"
+        description: str = "Test description"
+        categories: str = "Test category"
+
+    @classmethod
+    def split_list_to_txt_list(cls, usr_list: list[str], delimiter: str = " - ") -> str:
+        """Split list to multi-lines string.
+
+        Parameters
+        ----------
+            usr_list: List of string to concatenate
+            delimiter: A delimiter to use to start string. Defaults to None.
+
+        Returns
+        -------
+            str: Multi-lines string
+
+        """
+        return f"{delimiter}".join(f"{line}" for line in usr_list)
+
+    @classmethod
+    def convert_to_list(cls, result: TestResult) -> list[str]:
+        """
+        Convert a TestResult into a list of string for creating file content.
+
+        Args:
+        ----
+            results: A TestResult to convert into list.
+        """
+        message = cls.split_list_to_txt_list(result.messages) if len(result.messages) > 0 else ""
+        categories = cls.split_list_to_txt_list(result.categories) if len(result.categories) > 0 else "None"
+        return [
+            str(result.name),
+            result.test,
+            result.result,
+            message,
+            result.description,
+            categories,
+        ]
+
+    @classmethod
+    def generate(cls, results: ResultManager, csv_filename: pathlib.Path) -> None:
+        """Build CSV flle with tests results.
+
+        Args:
+        ----
+            results: A ResultManager instance.
+            csv_filename: File path where to save CSV data.
+        """
+        headers = [
+            cls.Headers.device,
+            cls.Headers.test_name,
+            cls.Headers.test_status,
+            cls.Headers.messages,
+            cls.Headers.description,
+            cls.Headers.categories,
+        ]
+
+        with pathlib.Path.open(csv_filename, "w", encoding="utf-8") as csvfile:
```

Comment:
this could probably fail and raise some OS error in some places right?

Reply:
[`Pathlib`](https://docs.python.org/3/library/pathlib.html) is taking care of host filesystem, so I don't think it will raise some errors.

Reply:
```
>>> p = pathlib.Path("/tmp/test_ro/ro_folder/toto.yml")
>>> with pathlib.Path.open(p, "w", encoding="utf-8")  as f:
...   f.write("blah")
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/guillaumemulocher/.pyenv/versions/3.10.4/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pathlib.py", line 1117, in open
    return self._accessor.open(self, mode, buffering, encoding, errors,
PermissionError: [Errno 13] Permission denied: '/tmp/test_ro/ro_folder/toto.yml'
```

not if it is a read only folder
we should catch OSError and raise a message indicating what goes wrong

Reply:
got it